client/heartbeatstop.go

```diff
@@ -0,0 +1,155 @@
+package client
+
+import (
+	"sync"
+	"time"
+
+	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+type heartbeatStop struct {
+	lastOk        time.Time
+	startupGrace  time.Time
+	allocInterval map[string]time.Duration
+	allocHookCh   chan *structs.Allocation
+	getRunner     func(string) (AllocRunner, error)
+	logger        hclog.InterceptLogger
+	shutdownCh    chan struct{}
+	lock          *sync.RWMutex
+}
+
+func newHeartbeatStop(
+	getRunner func(string) (AllocRunner, error),
+	timeout time.Duration,
+	logger hclog.InterceptLogger,
+	shutdownCh chan struct{}) *heartbeatStop {
+
+	h := &heartbeatStop{
+		startupGrace:  time.Now().Add(timeout),
+		allocInterval: make(map[string]time.Duration),
+		allocHookCh:   make(chan *structs.Allocation),
+		getRunner:     getRunner,
+		logger:        logger,
+		shutdownCh:    shutdownCh,
+		lock:          &sync.RWMutex{},
+	}
+
+	return h
+}
+
+// allocHook is called after (re)storing a new AllocRunner in the client. It registers the
+// allocation to be stopped if the taskgroup is configured appropriately
+func (h *heartbeatStop) allocHook(alloc *structs.Allocation) {
+	tg := allocTaskGroup(alloc)
+	if tg.StopAfterClientDisconnect != nil {
+		h.allocHookCh <- alloc
+	}
+}
+
+// shouldStop is called on a restored alloc to determine if lastOk is sufficiently in the
+// past that it should be prevented from restarting
+func (h *heartbeatStop) shouldStop(alloc *structs.Allocation) bool {
+	tg := allocTaskGroup(alloc)
+	if tg.StopAfterClientDisconnect != nil {
+		return h.shouldStopAfter(time.Now(), *tg.StopAfterClientDisconnect)
+	}
+	return false
+}
+
+func (h *heartbeatStop) shouldStopAfter(now time.Time, interval time.Duration) bool {
+	if h.lastOk.IsZero() {
```

Comment:
should this method also use the `getLastOk()` accessor in the two places we check `lastOk`?

Reply:
good catch, yup!